STM32/Libraries/UnitTesting/src/stubs.cpp

```diff
@@ -102,34 +107,63 @@ int calCompare(int noOfPorts, const CACalibration* catAr)
     return 0; // All good
 }
 
-int testCAProtocol()
+class TestCAProtocol
 {
-    CAProtocolCtx caProto = { 0 };
-    caProto.calibration = CAClibrationCb;
-    initCAProtocol(&caProto);
+public:
+    TestCAProtocol()
+    {
+        caProto.calibration = CAClibrationCb;
+        initCAProtocol(&caProto, testReader);
+        reset();
+    }
+    void reset() {
+        while(!testString.empty())
+            testString.pop();
+    }
 
-    inputCAProtocol(&caProto, "CAL 3,0.05,1.56\n");
-    if (calCompare(1, (CACalibration[]) {{3, 0.05, 1.56}} )) return __LINE__;
+    int testCalibration(const char* input, int noOfPorts, const CACalibration calAray[])
+    {
+        while(*input != 0) {
+            testString.push(*input);
+            input++;
+        }
+        inputCAProtocol(&caProto);
+        return calCompare(noOfPorts, calAray);
+    }
+
+private:
+    CAProtocolCtx caProto;;
+    static std::queue<uint8_t> testString;
+    static int testReader(uint8_t* rxBuf)
+    {
+        *rxBuf = testString.front();
+        testString.pop();
```

Comment:
In line 127 after push there is a input++. Should there be a input-- here?

Reply:
input is a pointer. It is pointer into a string  (array of chars) that is pushed into a queue.